File: Algorithms/Easy/4728_krosh_and_game/solution.dart
import 'dart:io';

void main() {
  solve();
}

void solve() {
  const int MAX_N = 100000;
  List<bool> dp = List<bool>.filled(MAX_N + 1, false);

  // Precompute perfect squares up to MAX_N
  List<int> perfectSquares = [];
  int i = 1;
  while (i * i <= MAX_N) {
    perfectSquares.add(i * i);
    i++;
  }

  for (int j = 1; j <= MAX_N; j++) {
    bool canWinFromJ = false;
    for (int square in perfectSquares) {
      if (j - square >= 0) {
        if (!dp[j - square]) {
          canWinFromJ = true;
          break;
        }
      } else {
        break;
      }
    }
    dp[j] = canWinFromJ;
  }

  int q = int.parse(stdin.readLineSync()!);
  List<String> results = [];
  for (int k = 0; k < q; k++) {
    int n = int.parse(stdin.readLineSync()!);
    results.add(dp[n] ? "1" : "0");
  }

  stdout.write(results.join("\n"));
}

File: Algorithms/Easy/4728_krosh_and_game/solution.py
# Solution for https://coderun.yandex.ru/problem/krosh-and-game
# Other solutions: https://github.com/Melodiz/CodeRun

def main():
    MAX_N = 100000
    dp = [False] * (MAX_N + 1)

    # Precompute perfect squares up to MAX_N
    perfect_squares = []
    i = 1
    while i * i <= MAX_N:
        perfect_squares.append(i * i)
        i += 1

    # dp[0] is False (base case)
    # dp[i] is True if there exists a move to a state j where dp[j] is False
    # dp[i] is False if all moves lead to a state j where dp[j] is True
    for i in range(1, MAX_N + 1):
        can_win_from_i = False
        for square in perfect_squares:
            if i - square >= 0:
                if not dp[i - square]:
                    can_win_from_i = True
                    break
        dp[i] = can_win_from_i

    q = int(input())
    for _ in range(q):
        n = int(input())
        print("1" if dp[n] else "0")


if __name__ == "__main__":
    main()
